lib/mysql/mysql.go

```diff
@@ -0,0 +1,651 @@
+const (
+	DEFAULT_TIMEOUT_SECS        = 3
+	DEFAULT_PORT                = 3306
+	DEFAULT_CLIENT_CAPABILITIES = CLIENT_SSL
+	DEFAULT_RESERVED_DATA_HEX   = "0000000000000000000000000000000000000000000000"
+)
+
+type Config struct {
+	// @TODO: Does it make sense to make Host/Port connection fields, so that Config can be shared across connections?
+	Host string
+	Port uint16
+
+	TLSConfig          *tls.Config
+	Timeout            time.Duration
+	ClientCapabilities uint32
+	MaxPacketSize      uint32
+	CharSet            byte
+	ReservedData       []byte
+}
+
+// Fill in a (possibly newly-created) Config instance with the default values
+func NewConfig(base *Config) *Config {
+	if base == nil {
+		base = &Config{}
+	}
+	if base.TLSConfig == nil {
+		// @TODO @FIXME Can this be pulled from a global zgrab config module?
+		base.TLSConfig = &tls.Config{InsecureSkipVerify: true}
+	}
+	if base.Port == 0 {
+		base.Port = DEFAULT_PORT
+	}
+	if base.Timeout == 0 {
+		base.Timeout = DEFAULT_TIMEOUT_SECS * time.Second
+	}
+	if base.ClientCapabilities == 0 {
+		base.ClientCapabilities = DEFAULT_CLIENT_CAPABILITIES
+	}
+	if base.ReservedData == nil {
+		bin, err := hex.DecodeString(DEFAULT_RESERVED_DATA_HEX)
+		if err != nil {
+			log.Fatalf("Invalid constant")
+		}
+		base.ReservedData = bin
+	}
+	return base
+}
+
+// Struct holding state for a single connection
+type Connection struct {
+	// Configuration for this connection
+	Config *Config
+
+	// Enum to track connection status
+	State ConnectionState
+	// TCP or TLS-wrapped Connection pointer (IsSecure will tell which)
+	Connection *net.Conn
+	// The sequence number used with the server to number packets
+	SequenceNumber uint8
+	// The "Handshake" packet sent by the server, holding flags used in future calls
+	Handshake *Handshake_Packet
+
+	// If this is true, the Connection is a TLS connection object and there should be a TLSHandshake log.
+	IsSecure bool
+
+	// List of MySQL packets received/sent
+	PacketLog []*PacketLogEntry
+	// The zgrab TLS handshake logs
+	TLSHandshake *tls.ServerHandshake
+}
+
+// Constructor, filling in defaults where needed
+func NewConnection(config *Config) *Connection {
+	return &Connection{
+		Config:         NewConfig(config),
+		State:          STATE_NOT_CONNECTED,
+		PacketLog:      nil,
+		Connection:     nil,
+		SequenceNumber: 0}
+}
+
+// Top-level interface for all packets
+type PacketInfo interface {
+	GetType() PacketType
+	GetDescription() string
+}
+
+// Most packets are read from the server; for packets that need to be sent, they need an encoding function
+type WritablePacket interface {
+	PacketInfo
+	EncodeBody() []byte
+}
+
+// Entry in the PacketLog. Raw is the base64-encoded body, Parsed is the parsed packet.
+// Either may be nil if there was an error reading/decoding the packet.
+type PacketLogEntry struct {
+	Length         uint32     `json:"length"`
+	SequenceNumber uint8      `json:"sequence_number"`
+	Raw            string     `json:"raw"`
```

Comment:
I'd add `omitempty` to this.

Reply:
My thinking was, if a packet log entry is created, it is saying there _was_ a packet — possibly with 0 length (I don’t think that is actually permitted, but it is conceivable). So I didn’t want to omit either empty strings or 0’s.